UserDetails/forms.py

```diff
@@ -52,12 +57,141 @@ class Meta:
+        # Check input data for correctness
+        if self.user is None and user is None:
+            raise ValueError("Field does not contain user and user was not given in method")
+        if user is not None and self.user is not None and self.user != user:
+            raise ValueError("Given user differs from field user")
+
+        if self.membership is not None:
+            return self.membership
+        if self.user is not None:
+            # If there was a user given, but the link was not found. Create a new link if allowed
+            if new_value:
+                return UserMembership(related_user=self.user, association=self.association)
+        else:
+            # user originally not given. Try to find the link
+            try:
+                return self.association.usermembership_set.get(related_user=user)
+            except ObjectDoesNotExist:
+                if new_value:
+                    return UserMembership(related_user=user, association=self.association)
+        return None
+
+
+class AssociationLinkForm(forms.Form):
+
+    def __init__(self, user, *args, **kwargs):
+        super(AssociationLinkForm, self).__init__(*args, **kwargs)
+
+        self.user = user
+
+        if user is None:
+            associations = Association.objects.filter(is_choosable=True)
+        else:
+            associations = Association.objects.filter(
+                Q(is_choosable=True) |
+                (Q(is_choosable=False) & Q(usermembership__related_user=user))) \
+                .distinct().order_by('slug')
+
+        # Get all associations and make a checkbox field
+        for association in associations:
+            field = AssociationLinkField(user, association)
+
+            self.fields[field.label] = field
+
+    def clean(self):
+        cleaned_data = super(AssociationLinkForm, self).clean()
+        has_association = False
+        for key, value in self.cleaned_data.items():
+            if value:
+                has_association = True
+
+        if not has_association:
+            raise ValidationError("At least one association needs to be chosen")
+
+        return cleaned_data
+
+    def save(self, user=None):
+        """
+        Saves the association links. Removes
+        :return:
+        """
+        if self.user is None and user is None:
+            raise ValueError("Both self.user and user are None")
+        if user is None:
+            user = self.user
+
+        for key, value in self.cleaned_data.items():
+            link = self.fields[key].get_membership_model(user, new_value=value)
+            if value:
+                if link.id is None:
+                    link.save()
+                elif link.get_verified_state() == False:
+                    # If user was rejected, and a new request is entered
+                    link.verified_on = None
+                    link.save()
+            else:
+                if link is not None and link.get_verified_state() != False:
```

Comment:
See comment above, you probably want `is not` instead of `!=`. Also, if you here do not allow removal of rejected user memberships, I don't see the need of a cooldown period at all.

Reply:
Ok. What happens in my approach is that if a member is denied and is still within the denial period the field becomes disabled. As a result the contents of this field aren't send along in a post request and can therefore not end up in this particular piece of code (although I don't know if this also occurs if you create a custom post request now that I think of it).
If a rejected membership ends here it can either be as False (i.e. it's not a membership) or True (the user wants to become a member).
- False: nothing should occur with the membership, otherwise you remove the rejection completely even if the user doesn't want to be a member
- True: The user was originally rejected, but still wants (/has become) a member and wants to request it again.
This particular piece of code deals with that the False case (186: if value:) where the membership should not be removed. The lines above 189-182. Takes care of the true case. The member was rejected, but wants a new request.
That said, I've shortened if link is not None to if link and... != False remains the same, as None != False results in True, which is what I want (along with True != False yielding True). I know it is a bit unorthodox.